File: Luca/app/src/main/java/de/culture4life/luca/ui/myluca/TestResultItem.java
package de.culture4life.luca.ui.myluca;

import android.content.Context;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;

import de.culture4life.luca.R;
import de.culture4life.luca.document.Document;
import de.culture4life.luca.util.TimeUtil;

public class TestResultItem extends MyLucaListItem {

    public static class TestProcedure {

        private final String name;
        private final String date;

        public TestProcedure(@NonNull String name, @NonNull String date) {
            this.name = name;
            this.date = date;
        }

        public String getName() {
            return name;
        }

        public String getDate() {
            return date;
        }

    }

    public TestResultItem(@NonNull Context context, @NonNull Document document) {
        super(TYPE_TEST_RESULT, document);

        this.title = getReadableTestType(context, document) + ": " + getReadableOutcome(context, document);
        this.provider = getReadableProvider(context, document.getProvider());
        if (document.isEudcc()) {
            this.barcode = generateQrCode(document.getEncodedData()).blockingGet();
        }
        this.color = getColor(context, document);
        this.deleteButtonText = context.getString(R.string.delete_test_action);
        this.imageResource = document.isVerified() ? R.drawable.ic_verified : R.drawable.ic_warning_triangle_orange;

        String time = context.getString(R.string.document_result_time, TimeUtil.getReadableTime(context, document.getResultTimestamp()));
        addTopContent(context.getString(R.string.document_issued), time);

        String duration = TimeUtil.getReadableDurationWithPlural(TimeUtil.getCurrentMillis() - document.getResultTimestamp(), context).blockingGet();
        addTopContent(context.getString(R.string.document_created_before), duration);

        addCollapsedContent(context.getString(R.string.document_lab_issuer), document.getLabName());
        addCollapsedContent(context.getString(R.string.document_lab_tester), document.getLabDoctorName());
    }

    public static String getReadableTestType(@NonNull Context context, @NonNull Document document) {
        switch (document.getType()) {
            case Document.TYPE_FAST: {
                return context.getString(R.string.document_type_fast);
            }
            case Document.TYPE_PCR: {
                return context.getString(R.string.document_type_pcr);
            }
            default: {
                return context.getString(R.string.document_type_unknown);
            }
        }
    }

    public static String getReadableOutcome(@NonNull Context context, @NonNull Document document) {
        switch (document.getOutcome()) {
            case Document.OUTCOME_POSITIVE: {
                return context.getString(R.string.document_outcome_positive);
            }
            case Document.OUTCOME_NEGATIVE: {
                return context.getString(R.string.document_outcome_negative);
            }
            default: {
                return context.getString(R.string.document_outcome_unknown);
            }
        }
    }

    public static String getReadableResult(@NonNull Context context, @NonNull Document document) {
        return context.getString(R.string.document_outcome, getReadableOutcome(context, document));
    }

    @ColorInt
    private static int getColor(@NonNull Context context, @NonNull Document document) {
        switch (document.getOutcome()) {
            case Document.OUTCOME_POSITIVE: {
                if (document.isValidRecovery()) {
                    return ContextCompat.getColor(context, R.color.document_outcome_partially_vaccinated);
                } else {
                    return ContextCompat.getColor(context, R.color.document_outcome_positive);
                }
            }
            case Document.OUTCOME_NEGATIVE: {
                return ContextCompat.getColor(context, R.color.document_outcome_negative);
            }
            default: {
                return ContextCompat.getColor(context, R.color.document_outcome_unknown);
            }
        }
    }

}

File: Luca/app/src/main/java/de/culture4life/luca/ui/myluca/TestResultItem.kt
package de.culture4life.luca.ui.myluca

import android.content.Context
import androidx.core.content.ContextCompat
import de.culture4life.luca.R
import de.culture4life.luca.document.Document
import de.culture4life.luca.document.Document.OUTCOME_NEGATIVE
import de.culture4life.luca.document.Document.OUTCOME_POSITIVE
import de.culture4life.luca.util.TimeUtil
import de.culture4life.luca.util.TimeUtil.getReadableTime
import de.culture4life.luca.util.toDateTime
import org.joda.time.LocalDate

open class TestResultItem(context: Context, document: Document) : MyLucaListItem(TYPE_TEST_RESULT, document) {

    init {
        title = getReadableTestType(context, document) + ": " + getReadableOutcome(context, document)
        deleteButtonText = context.getString(R.string.delete_test_action)
        imageResource = if (document.isVerified) R.drawable.ic_verified else R.drawable.ic_warning_triangle_orange
        provider = getReadableProvider(context, document.provider)
        if (document.isEudcc) barcode = generateQrCode(document.encodedData).blockingGet()

        setupColor(context, document)
        setupTopContent(context)
        setupCollapsableContent(context)
    }

    private fun setupTopContent(context: Context) {
        topContent.clear()
        addTopContent(DynamicContent("${document.firstName} ${document.lastName}", ""))
        val time = context.getString(R.string.document_result_time, getReadableTime(context, document.resultTimestamp))
        addTopContent(DynamicContent(context.getString(R.string.document_issued), time))

        val resultDate = document.resultTimestamp.toDateTime()
        val duration = TimeUtil.getReadableDateTimeDifference(context, resultDate)
        addTopContent(
            DynamicContent(
                context.getString(R.string.testing_done_before_label),
                duration,
                if (resultDate.toLocalDate().equals(LocalDate.now())) R.drawable.ic_rocket else null
            )
        )
    }

    private fun setupCollapsableContent(context: Context) {
        addCollapsedContent(DynamicContent(context.getString(R.string.document_lab_issuer), document.labName))
        addCollapsedContent(DynamicContent(context.getString(R.string.document_lab_tester), document.labDoctorName))
    }

    private fun setupColor(context: Context, document: Document) {
        color = ContextCompat.getColor(
            context,
            when (document.outcome) {
                OUTCOME_POSITIVE -> if (document.isValidRecovery) R.color.document_outcome_partially_vaccinated else R.color.document_outcome_positive
                OUTCOME_NEGATIVE -> R.color.document_outcome_negative
                else -> R.color.document_outcome_unknown
            }
        )
    }

    data class TestProcedure(val name: String, val date: String)

    companion object {
        @JvmStatic
        fun getReadableTestType(context: Context, document: Document) = when (document.type) {
            Document.TYPE_FAST -> context.getString(R.string.document_type_fast)
            Document.TYPE_PCR -> context.getString(R.string.document_type_pcr)
            else -> context.getString(R.string.document_type_unknown)
        }

        @JvmStatic
        fun getReadableOutcome(context: Context, document: Document) = when (document.outcome) {
            OUTCOME_POSITIVE -> context.getString(R.string.document_outcome_positive)
            OUTCOME_NEGATIVE -> context.getString(R.string.document_outcome_negative)
            else -> context.getString(R.string.document_outcome_unknown)
        }

        @JvmStatic
        fun getReadableResult(context: Context, document: Document) =
            context.getString(R.string.document_outcome, getReadableOutcome(context, document))
    }
}
